manager/controllers/app/select_modules.go

```diff
@@ -116,51 +115,81 @@ func (p *PlotterGenerator) validSolutions(item *DataInfo, solutions []Solution,
 	return validPaths
 }
 
-//nolint:gocyclo
-func (p *PlotterGenerator) validate(item *DataInfo, solution Solution, application *app.FybrikApplication) bool {
-	// start from data source, check supported actions and cluster restrictions
+// if new storage should be located, check the requirements:
+// where storage can be allocated
+// what additional actions to perform
+func (p *PlotterGenerator) validateStorageRequirements(item *DataInfo, application *app.FybrikApplication, element *ResolvedEdge) bool {
 	appContext := ApplicationContext{Application: application, Log: p.Log}
-	requiredActions := item.Actions
-	for _, element := range solution.DataPath {
-		element.Actions = []taxonomy.Action{}
-		moduleCapability := element.Module.Spec.Capabilities[element.CapabilityIndex]
-		if !element.Edge.Sink.Virtual && (item.Context.Flow != taxonomy.WriteFlow ||
-			item.Context.Requirements.FlowParams.IsNewDataSet) {
-			// storage is required, plus more actions on copy may be needed
-			isAccountFound := false
-			for accountInd := range p.StorageAccounts {
-				// validate restrictions
-				if !p.validateRestrictions(item.Configuration.ConfigDecisions[moduleCapability.Capability].
-					DeploymentRestrictions.StorageAccounts, &p.StorageAccounts[accountInd].Spec,
-					p.StorageAccounts[accountInd].Name) {
-					p.Log.Debug().Str(logging.DATASETID, item.Context.DataSetID).Msgf("storage account %s does not match the requirements",
-						p.StorageAccounts[accountInd].Name)
-					continue
-				}
-				// query the policy manager whether WRITE operation is allowed
-				operation := new(policymanager.RequestAction)
-				operation.ActionType = taxonomy.WriteFlow
+	var err error
+	var actions []taxonomy.Action
+	operation := &policymanager.RequestAction{}
+	operation.ActionType = taxonomy.WriteFlow
+
+	if item.Context.Flow == taxonomy.WriteFlow && !item.Context.Requirements.FlowParams.IsNewDataSet {
```

Comment:
No. This function is common to all cases where data should be written.
The cases are:
- Read with implicit copy
- Copy (ingest)
- Write where the destination is known (update)
- Write where the destination is unknown.
Except for the third case (update), Fybrik should select a storage account out of the existing and allocate a storage. 
The condition in line 128 checks the update flow (write to an existing dataset)

Reply:
Rewritten this function. There is no point to check for governance actions on each constructed path for an update flow since the information is known beforehand.